azure/durable_functions/decorators/durable_app.py

```diff
@@ -0,0 +1,264 @@
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License.
+from .triggers import OrchestrationTrigger, ActivityTrigger, EntityTrigger,\
+    EntityClient, OrchestrationClient, DurableClient
+from typing import Callable, Optional
+from azure.durable_functions.entity import Entity
+from azure.durable_functions.orchestrator import Orchestrator
+from typing import Union
+from azure.functions import FunctionRegister, TriggerApi, BindingApi, AuthLevel
+
+
+class DFApp(FunctionRegister, TriggerApi, BindingApi):
+    """Durable Functions (DF) app.
+
+    Exports the decorators required to register DF Function-types.
+    """
+
+    def __init__(self,
+                 http_auth_level: Union[AuthLevel, str] = AuthLevel.FUNCTION):
+        """Instantiate a Durable Functions app with which to register Functions.
+
+        Parameters
+        ----------
+        http_auth_level: Union[AuthLevel, str]
+            Authorization level required for Function invocation.
+            Defaults to AuthLevel.Function.
+
+        Returns
+        -------
+        DFApp
+            New instance of a Durable Functions app
+        """
+        super().__init__(auth_level=http_auth_level)
+
+    def _configure_entity_callable(self, wrap) -> Callable:
+        """Obtain decorator to construct an Entity class from a user-defined Function.
+
+        In the old programming model, this decorator's logic was unavoidable boilerplate
+        in user-code. Now, this is handled internally by the framework.
+
+        Parameters
+        ----------
+        wrap: Callable
+            The next decorator to be applied.
+
+        Returns
+        -------
+        Callable
+            The function to construct an Entity class from the user-defined Function,
+            wrapped by the next decorator in the sequence.
+        """
+        def decorator(entity_func):
+            # Construct an entity based on the end-user code
+            handle = Entity.create(entity_func)
+
+            # invoke next decorator, with the Entity as input
+            handle.__name__ = entity_func.__name__
+            return wrap(handle)
+
+        return decorator
```

Comment:
A lot of methods in this PR look like this one. This is because this defines a decorator (i.e a Python annotation), and decorators are usually implemented in this pattern.

In this case, the `_` prefix at the beginning of the method tells us this is an internal decorator, so this is not what PyStein users will interface with. This is used internally